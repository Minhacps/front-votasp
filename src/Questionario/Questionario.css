.questionario__container {
  display: flex;
  flex-direction: column;
  /*height: calc(100vh - 100px);*/
  padding: 20px;
}

.questionario__container--loading {
  opacity: 0.4;
}

.questionario__actions-container {
  display: flex;
  justify-content: space-around;
  flex-shrink: 0;
  margin-top: 30px;
}

.questionario__aviso-candidato {
  color: #727273;
  font-size: 12px;
  margin: 0 0 10px 0;
}

.justification-field {
  margin-top: 20px;
}

.justification-field textarea {
  height: 94px;
}

.questions_box {
  width: 100%;
  margin: 0 auto;
}

@media screen and (min-width: 800px) {
  .questions_box {
    width: 50%;
  }
}
