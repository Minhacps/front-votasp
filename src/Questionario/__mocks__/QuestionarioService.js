export const saveVoterAnswer = jest.fn(() => Promise.resolve());

export const getCurrentUser = jest.fn(() =>
  Promise.resolve({
    data: () => ({
      role: 'candidate'
    })
  })
);

export const watchAnswers = jest.fn(() => ({
  onSnapshot: () => Promise.resolve()
}));

export const watchAnswerJustification = jest.fn(() => ({
  onSnapshot: () => Promise.resolve()
}));
