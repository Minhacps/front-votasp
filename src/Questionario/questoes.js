const questoes = [
  {
    "id": 1,
    "question": "Liberação do porte de armas no Brasil."
  },
  {
    "id": 2,
    "question": "Regulamentação dos meios de comunicação."
  },
  {
    "id": 3,
    "question": "O governo tem um papel essencial na regulação dos mercados para garantir os direitos dos consumidores e consumidoras."
  },
  {
    "id": 4,
    "question": "Maior carga tributária para empresas estrangeiras com o objetivo de beneficiar empresas brasileiras."
  },
  {
    "id": 5,
    "question": "Garantia pelo governo de aposentadoria para todas e todos os brasileiros."
  },
  {
    "id": 6,
    "question": "Saúde pública de qualidade e gratuita para toda a população."
  },
  {
    "id": 7,
    "question": "Isenção de impostos de instituições religiosas."
  },
  {
    "id": 8,
    "question": "Legalização do consumo da maconha."
  },
  {
    "id": 9,
    "question": "Adoção de cotas raciais nas universidades públicas."
  },
  {
    "id": 10,
    "question": "Garantia de mesmos direitos dos brasileiros(as) para imigrantes."
  },
  {
    "id": 11,
    "question": "Os ricos deveriam pagar mais impostos."
  },
  {
    "id": 12,
    "question": "Legalização do aborto para todos os casos."
  },
  {
    "id": 13,
    "question": "Pena de morte."
  },
  {
    "id": 14,
    "question": "Adoção de crianças por casais do mesmo sexo."
  },
  {
    "id": 15,
    "question": "Crescimento econômico é mais importante que igualdade social."
  },
  {
    "id": 16,
    "question": "Reforma agrária."
  },
  {
    "id": 17,
    "question": "Debate de gênero nas escolas."
  },
  {
    "id": 18,
    "question": "Redução da maioridade penal para 16 anos."
  },
  {
    "id": 19,
    "question": "Definição exclusiva de família como homem mulher e filhos."
  },  
  {
    "id": 20,
    "question": "Privatização de serviços públicos como educação, saúde e cultura."
  },
  {
    "id": 21,
    "question": "Campanhas eleitorais financiadas unicamente com dinheiro público."
  },
  {
    "id": 22,
    "question": "Construção de moradias populares na periferia mesmo antes de garantir atendimento básico como centros de saúde e escolas."
  },
  {
    "id": 23,
    "question": "Grandes empreendimentos podem iniciar suas obras mesmo antes de conseguirem licenciamento ambiental."
  },
  {
    "id": 24,
    "question": "A conciliação é uma medida útil para ajudar nos casos de violência doméstica e familiar."
  },
  {
    "id": 25,
    "question": "Interferência do estado na economia."
  },
  {
    "id": 26,
    "question": "Leis para garantir a igualdade de salários entre homens e mulheres."
  },
  {
    "id": 27,
    "question": "Intervenção militar para combater a criminalidade."
  },
  {
    "id": 28,
    "question": "Privatização da previdência social."
  },
  {
    "id": 29,
    "question": "As Universidades Públicas deveriam ser paga por quem pode pagar."
  },
  {
    "id": 30,
    "question": "A reforma trabalhista de 2017 modernizou as relações de trabalho do Brasil."
  },
  {
    "id": 31,
    "question": "Construções de barragens para solucionar a crise hídrica."
  },
  {
    "id": 32,
    "question": "Fim do auxílio moradia para juízes."
  },
  {
    "id": 33,
    "question": "Privatização completa da Petrobrás."
  },
  {
    "id": 34,
    "question": "Desapropriação de prédios ociosos para moradia popular."
  },  
  {
    "id": 35,
    "question": "O Brasil é um país racista."
  },
  {
    "id": 36,
    "question": "Criminalização da homofobia no Brasil."
  },
  {
    "id": 37,
    "question": "Em briga de marido e mulher ninguém mete a colher."
  },
  {
    "id": 38,
    "question": "O financimento empresarial de campanhas políticas deixa o candidato com rabo preso."
  },
  {
    "id": 39,
    "question": "Fim do foro privilegiado."
  },  
  {
    "id": 40,
    "question": "Redução da verba de gabinete para parlamentares."
  }
]

export default questoes;
