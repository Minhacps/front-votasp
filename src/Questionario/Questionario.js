import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import get from 'lodash/get';

import PageLayout from '../components/PageLayout/PageLayout';
import Pergunta from '../components/Pergunta/Pergunta';
import QuestionsMenu from '../components/QuestionsMenu/QuestionsMenu';
import questoes from './questoes';
import store from '../redux/store';
import { INDIFERENTE } from '../constants/respostas';
import {
  saveAnswer,
  saveJustification,
  watchAnswers,
  getCurrentUser,
  watchAnswerJustification
} from './QuestionarioService';
import { storeQuestionario } from '../redux/modules/questionario';

import './Questionario.css';

export class RawQuestionario extends Component {
  static propTypes = {
    history: PropTypes.object,
    location: PropTypes.object,
    match: PropTypes.object,
    questionario: PropTypes.shape({
      currentQuestion: PropTypes.number
    })
  };

  state = {
    isAnswering: false,
    userAnswers: [],
    userCollection: undefined,
    isCandidate: false,
    currentJustification: '',
    candidateJustifications: []
  };

  componentDidMount() {
    const { question } = this.props.match.params;

    store.dispatch(
      storeQuestionario({
        currentQuestion: parseInt(question - 1)
      })
    );

    getCurrentUser().then(this.saveUser);
  }

  componentDidUpdate(prevProps) {
    const lastQuestion = prevProps.match.params.question;
    const { question } = this.props.match.params;

    if (question !== lastQuestion) {
      store.dispatch(
        storeQuestionario({
          currentQuestion: parseInt(question - 1)
        })
      );

      this.setCurrentJustification();
    }
  }

  setCurrentJustification = () => {
    const { isCandidate, candidateJustifications, currentJustification } = this.state;

    if (!isCandidate) {
      return;
    }

    const { currentQuestion } = this.props.questionario;

    const [storedJustification] = candidateJustifications.filter(
      answer => answer.id == currentQuestion + 1
    );

    const questionJustification = get(storedJustification, 'justification') || '';

    this.setState({
      currentJustification: questionJustification
    });
  };

  saveUser = doc => {
    const user = doc.data();
    const isCandidate = user.role === 'candidate';
    const userCollection = isCandidate ? 'candidate_answers' : 'voter_answers';

    watchAnswers(userCollection).onSnapshot(this.storeAnswers);
    watchAnswerJustification().onSnapshot(this.storeJustification);

    this.setState({
      userCollection,
      isCandidate
    });
  };

  storeAnswers = snapshot => {
    const data = snapshot.data();
    const userAnswers = data
      ? Object.keys(data).map(answerKey => ({
        id: answerKey,
        answer: data[answerKey]
      }))
      : [];

    this.setState({
      userAnswers
    });
  };

  storeJustification = snapshot => {
    const data = snapshot.data();
    const candidateJustifications = data
      ? Object.keys(data).map(answerKey => ({
        id: answerKey,
        justification: data[answerKey]
      }))
      : [];

    this.setState(
      {
        candidateJustifications
      },
      this.setCurrentJustification
    );
  };

  pularQuestao = () => {
    const { isCandidate } = this.state;
    if (isCandidate) return;

    return this.saveAnswer(INDIFERENTE.value).then(() => {
      this.proximaQuestao();
    });
  };

  responderQuestao = event => {
    const { isCandidate } = this.state;

    this.saveAnswer(event.target.value).then(() => {
      if (isCandidate) return;

      this.proximaQuestao();
    });
  };

  saveAnswer = answerValue => {
    const { userCollection, currentJustification } = this.state;
    const { questionario } = this.props;
    const questionId = questionario.currentQuestion + 1;
    const answer = {
      [questionId]: answerValue
    };

    const justification = {
      [questionId]: currentJustification
    };

    if (!this.state.isCandidate) {
      this.setState({
        isAnswering: true
      });
    }

    saveJustification(justification);

    return saveAnswer(answer, userCollection);
  };

  proximaQuestao = () => {
    const { questionario } = this.props;
    const { isCandidate } = this.state;

    if (isCandidate && questionario.currentQuestion === questoes.length - 1) {
      this.props.history.push('/questionario-finalizado');
      return;
    }

    if (questionario.currentQuestion === questoes.length - 1) {
      this.props.history.push('/app/ranking');
      return;
    }

    store.dispatch(
      storeQuestionario({
        currentQuestion: questionario.currentQuestion + 1
      })
    );

    this.props.history.push(`/app/questionario/${questionario.currentQuestion + 2}`);

    this.setState({
      isAnswering: false
    });
  };

  updateJustification = event => {
    this.setState({
      currentJustification: event.target.value
    });
  };

  saveCandidateAnswer = currentAnswer => {
    if (!currentAnswer) {
      return;
    }

    this.setState({
      isAnswering: true
    });

    this.saveAnswer(currentAnswer.answer).then(() => {
      this.proximaQuestao();
    });
  };

  render() {
    const { isAnswering, userAnswers, isCandidate, currentJustification } = this.state;
    const { questionario, history } = this.props;

    const { currentQuestion } = questionario;
    const [currentAnswer] = userAnswers.filter(answer => answer.id == currentQuestion + 1);

    return (
      <PageLayout>
        <div className={'questions_box'}>
          <QuestionsMenu
            userAnswers={userAnswers}
            questionsArray={questoes}
            currentQuestion={currentQuestion}
            history={history}
          />
          <div
            className={classnames('questionario__container', {
              'questionario__container--loading': isAnswering
            })}
          >
            {isCandidate && (
              <p className="questionario__aviso-candidato">
                VocÃª como candidato(a) precisa responder as 40 perguntas para se tornar apto(a) ao
                match.
              </p>
            )}

            {questoes.length && (
              <Pergunta
                pergunta={questoes[currentQuestion]}
                responderQuestao={this.responderQuestao}
                isAnswering={isAnswering}
                userAnswer={currentAnswer ? currentAnswer.answer : undefined}
              />
            )}

            {isCandidate && (
              <div className="justification-field field-wrapper">
                <label htmlFor="justification">
                  Justificativa <small>(opcional)</small>
                </label>
                <textarea
                  value={currentJustification}
                  onChange={this.updateJustification}
                  className="input"
                  name="justification"
                  id="justification"
                  maxLength={500}
                />
              </div>
            )}

            <div className="questionario__actions-container">
              {!isCandidate && (
                <button
                  onClick={this.pularQuestao}
                  className="btn btn-light"
                  disabled={isAnswering}
                >
                  Pular
                </button>
              )}

              {!isCandidate && (
                <Link
                  to="/app/ranking"
                  className="btn btn-light"
                  disabled={isAnswering || userAnswers.length < 20}
                >
                  Calcular afinidade
                </Link>
              )}

              {isCandidate && (
                <button
                  onClick={() => this.saveCandidateAnswer(currentAnswer)}
                  className="btn btn-primary"
                  disabled={isAnswering}
                >
                  Responder
                </button>
              )}
            </div>
          </div>
        </div>
      </PageLayout>
    );
  }
}

const mapStateToProps = ({ questionario }) => ({
  questionario
});

export default connect(mapStateToProps)(RawQuestionario);
