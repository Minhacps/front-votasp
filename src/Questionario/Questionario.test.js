import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import QuestionarioService from './QuestionarioService';
import { RawQuestionario } from './Questionario';

jest.mock('./QuestionarioService');

const defaultProps = {
  perguntas: [
    {
      id: 1,
      question: 'Legalização do aborto para todos os casos.'
    },
    {
      id: 2,
      question: 'Adoção de crianças por casais do mesmo sexo.'
    }
  ],
  match: {
    params: {}
  },
  history: {
    push: () => {}
  },
  questionario: {}
};

describe('<Questionario />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const questionario = shallow(<RawQuestionario {...defaultProps} />);

    expect(toJson(questionario)).toMatchSnapshot();
  });
});
