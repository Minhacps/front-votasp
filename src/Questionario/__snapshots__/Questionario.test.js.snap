// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Questionario /> should match snapshot 1`] = `
<PageLayout>
  <div
    className="questions_box"
  >
    <QuestionsMenu
      history={
        Object {
          "push": [Function],
        }
      }
      questionsArray={
        Array [
          Object {
            "id": 1,
            "question": "Liberação do porte de armas no Brasil.",
          },
          Object {
            "id": 2,
            "question": "Regulamentação dos meios de comunicação.",
          },
          Object {
            "id": 3,
            "question": "O governo tem um papel essencial na regulação dos mercados para garantir os direitos dos consumidores e consumidoras.",
          },
          Object {
            "id": 4,
            "question": "Maior carga tributária para empresas estrangeiras com o objetivo de beneficiar empresas brasileiras.",
          },
          Object {
            "id": 5,
            "question": "Garantia pelo governo de aposentadoria para todas e todos os brasileiros.",
          },
          Object {
            "id": 6,
            "question": "Saúde pública de qualidade e gratuita para toda a população.",
          },
          Object {
            "id": 7,
            "question": "Isenção de impostos de instituições religiosas.",
          },
          Object {
            "id": 8,
            "question": "Legalização do consumo da maconha.",
          },
          Object {
            "id": 9,
            "question": "Adoção de cotas raciais nas universidades públicas.",
          },
          Object {
            "id": 10,
            "question": "Garantia de mesmos direitos dos brasileiros(as) para imigrantes.",
          },
          Object {
            "id": 11,
            "question": "Os ricos deveriam pagar mais impostos.",
          },
          Object {
            "id": 12,
            "question": "Legalização do aborto para todos os casos.",
          },
          Object {
            "id": 13,
            "question": "Pena de morte.",
          },
          Object {
            "id": 14,
            "question": "Adoção de crianças por casais do mesmo sexo.",
          },
          Object {
            "id": 15,
            "question": "Crescimento econômico é mais importante que igualdade social.",
          },
          Object {
            "id": 16,
            "question": "Reforma agrária.",
          },
          Object {
            "id": 17,
            "question": "Debate de gênero nas escolas.",
          },
          Object {
            "id": 18,
            "question": "Redução da maioridade penal para 16 anos.",
          },
          Object {
            "id": 19,
            "question": "Definição exclusiva de família como homem mulher e filhos.",
          },
          Object {
            "id": 20,
            "question": "Privatização de serviços públicos como educação, saúde e cultura.",
          },
          Object {
            "id": 21,
            "question": "Campanhas eleitorais financiadas unicamente com dinheiro público.",
          },
          Object {
            "id": 22,
            "question": "Construção de moradias populares na periferia mesmo antes de garantir atendimento básico como centros de saúde e escolas.",
          },
          Object {
            "id": 23,
            "question": "Grandes empreendimentos podem iniciar suas obras mesmo antes de conseguirem licenciamento ambiental.",
          },
          Object {
            "id": 24,
            "question": "A conciliação é uma medida útil para ajudar nos casos de violência doméstica e familiar.",
          },
          Object {
            "id": 25,
            "question": "Interferência do estado na economia.",
          },
          Object {
            "id": 26,
            "question": "Leis para garantir a igualdade de salários entre homens e mulheres.",
          },
          Object {
            "id": 27,
            "question": "Intervenção militar para combater a criminalidade.",
          },
          Object {
            "id": 28,
            "question": "Privatização da previdência social.",
          },
          Object {
            "id": 29,
            "question": "As Universidades Públicas deveriam ser pagas por quem pode pagar.",
          },
          Object {
            "id": 30,
            "question": "A reforma trabalhista de 2017 modernizou as relações de trabalho do Brasil.",
          },
          Object {
            "id": 31,
            "question": "Construções de barragens para solucionar a crise hídrica.",
          },
          Object {
            "id": 32,
            "question": "Fim do auxílio moradia para juízes.",
          },
          Object {
            "id": 33,
            "question": "Privatização completa da Petrobrás.",
          },
          Object {
            "id": 34,
            "question": "Desapropriação de prédios ociosos para moradia popular.",
          },
          Object {
            "id": 35,
            "question": "O Brasil é um país racista.",
          },
          Object {
            "id": 36,
            "question": "Criminalização da homofobia no Brasil.",
          },
          Object {
            "id": 37,
            "question": "Em briga de marido e mulher ninguém mete a colher.",
          },
          Object {
            "id": 38,
            "question": "O financimento empresarial de campanhas políticas deixa o candidato com rabo preso.",
          },
          Object {
            "id": 39,
            "question": "Fim do foro privilegiado.",
          },
          Object {
            "id": 40,
            "question": "Redução da verba de gabinete para parlamentares.",
          },
        ]
      }
      userAnswers={Array []}
    />
    <div
      className="questionario__container"
    >
      <Pergunta
        isAnswering={false}
        responderQuestao={[Function]}
      />
      <div
        className="questionario__actions-container"
      >
        <button
          className="btn btn-light"
          disabled={false}
          onClick={[Function]}
        >
          Pular
        </button>
        <Link
          className="btn btn-light"
          disabled={true}
          replace={false}
          to="/app/ranking"
        >
          Calcular afinidade
        </Link>
      </div>
    </div>
  </div>
</PageLayout>
`;
